libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+            
+            __JavaJNISetStaticFieldResult(t_target_class,
+                                          static_cast<jfieldID>(p_method_id),
+                                          &t_params[0],
+                                          t_field_type);
+        }
+    }
+    
+    return false;
+}
+
+bool MCJavaPrivateObjectDescribe(MCValueRef p_value, MCStringRef &r_desc)
+{
+    MCJavaObjectRef t_obj = static_cast<MCJavaObjectRef>(p_value);
+    
+    MCAutoStringRef t_class_name;
+    if (!MCJavaPrivateGetJObjectClassName(t_obj, &t_class_name))
+        return false;
+    
+    return MCStringFormat(r_desc, "<java: %@>", *t_class_name);
+}
+
+bool MCJavaPrivateConvertStringRefToJString(MCStringRef p_string, MCJavaObjectRef &r_object)
+{
+    jstring t_string;
+    if (!__MCJavaStringToJString(p_string, t_string))
+        return false;
+    
+    return MCJavaObjectCreateGlobalRef(t_string, r_object);
+}
+
+bool MCJavaPrivateConvertJStringToStringRef(MCJavaObjectRef p_object, MCStringRef &r_string)
+{
+    auto t_string = static_cast<jstring>(MCJavaObjectGetObject(p_object));
+    return __MCJavaStringFromJString(t_string, r_string);
+}
+
+bool MCJavaPrivateConvertDataRefToJByteArray(MCDataRef p_data, MCJavaObjectRef &r_object)
+{
+    jbyteArray t_array;
+    if (!__MCJavaDataToJByteArray(p_data, t_array))
+        return false;
+    
+    return MCJavaObjectCreateGlobalRef(t_array, r_object);
+}
+
+bool MCJavaPrivateConvertJByteArrayToDataRef(MCJavaObjectRef p_object, MCDataRef &r_data)
+{
+    auto t_data = static_cast<jbyteArray>(MCJavaObjectGetObject(p_object));
+    return __MCJavaDataFromJByteArray(t_data, r_data);
+}
+
+bool MCJavaPrivateGetJObjectClassName(MCJavaObjectRef p_object, MCStringRef &r_name)
+{
+    jobject t_object = static_cast<jobject>(MCJavaObjectGetObject(p_object));
+    jstring t_class_name = MCJavaGetJObjectClassName(t_object);
+    return __MCJavaStringFromJString(t_class_name, r_name);
+}
+
+void* MCJavaPrivateGetMethodId(MCNameRef p_class_name, MCStringRef p_method_name, MCStringRef p_arguments, MCStringRef p_return, int p_call_type)
+{
+    MCAutoStringRef t_class_path;
+    if (!MCJavaClassNameToPathString(p_class_name, &t_class_path))
+        return nullptr;
+    
+    MCAutoStringRefAsCString t_class_cstring, t_method_cstring, t_return_cstring;
+    t_class_cstring . Lock(*t_class_path);
+    t_method_cstring . Lock(p_method_name);
+    t_return_cstring . Lock(p_return);
+    
+    MCJavaDoAttachCurrentThread();
+    
+    jclass t_java_class = s_env->FindClass(*t_class_cstring);
+    
+    void *t_id = nullptr;
+    if (t_java_class != nullptr)
+    {
+        switch (p_call_type)
+        {
+            case MCJavaCallTypeInstance:
+            case MCJavaCallTypeNonVirtual:
+            case MCJavaCallTypeStatic:
+            {
+                MCAutoStringRef t_sig;
+                if (!MCStringFormat(&t_sig, "%@%@", p_arguments, p_return))
+                    return nullptr;
```

Comment:
`MCStringCreateWithStrings()` is a more efficient way to do this than `MCStringFormat()`.